csrc/scheduler/matmul.cpp

```diff
@@ -966,18 +975,24 @@ void scheduleMatmul(Fusion* fusion, const MatmulParams& params) {
   //   and needs more configurability.
   // ------------------------------------------------------------------
 
-  acr->setAllocationDomain(acr->getLeafDomain(), true);
-  bcr->setAllocationDomain(bcr->getLeafDomain(), true);
-  mma_utils::WarpMmaSwizzler::scheduleLdMatrix(acr, MmaOperand::A);
-  mma_utils::WarpMmaSwizzler::scheduleLdMatrix(bcr, MmaOperand::B);
-
-  //  -5  -4   -3   -2   -1          or          -5  -4   -3   -2   -1
-  //[8mi, 4k, 2ko, 2mo, 2ki]                   [8ni, 4k, 2ko, 1no, 2ki]
-  for (auto tv : {ab, bb}) {
-    tv->merge(-5);
-    tv->axis(-4)->parallelize(ParallelType::TIDx);
+  if (acr != ab) {
+    //  -5  -4   -3   -2   -1
+    //[8mi, 4k, 2ko, 2mo, 2ki]
+    acr->setAllocationDomain(acr->getLeafDomain(), true);
+    mma_utils::WarpMmaSwizzler::scheduleLdMatrix(acr, MmaOperand::A);
+    ab->merge(-5);
+    ab->axis(-4)->parallelize(ParallelType::TIDx);
+    propagate_mma_input_schedule_to(acr, nullptr);
+  }
+  if (bcr != bb) {
+    //   -5  -4   -3   -2   -1
+    // [8ni, 4k, 2ko, 1no, 2ki]
+    bcr->setAllocationDomain(bcr->getLeafDomain(), true);
+    mma_utils::WarpMmaSwizzler::scheduleLdMatrix(bcr, MmaOperand::B);
+    bb->merge(-5);
+    bb->axis(-4)->parallelize(ParallelType::TIDx);
+    propagate_mma_input_schedule_to(nullptr, bcr);
```

Comment:
> My understanding is that `acr` and `ab` would be swizzled already.

It is ambiguous to say "swizzled already". Basically, for Turing, Ampere, and Hopper matmul's register operand, there are **two** different swizzles to consider: the first (`scheduleOperandRead`) is the memory layout about which thread's which register get which data, an example layout can be found [here](https://docs.nvidia.com/cuda/parallel-thread-execution/_images/mma-16816-A-f16.png); the second (`scheduleLdMatrix`) is about which thread should use which index to call `ldmatrix`, this is shown [here](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#warp-level-matrix-instructions-ldmatrix). So it is really important to be clear about which "swizzle" we are talking about (unfortunately people just tend to overload the word "swizzle" too much).

The allocation domain for both mma inputs and ldmatrix outputs should be scheduled like the first swizzle above. For ldmatrix outputs, its leaf domain should also be scheduled like the second swizzle, and the transformations of the second swizzle should be stacked on top of the first swizzle. Due to this nature of the swizzle, currently `applyMmaSwizzle` does two things: schedule swizzle 1, if the tensor is ldmatrix output, then also schedule swizzle 2.

In the above code, we already called `ab->applyMmaSwizzle(MmaOperand::A)`, for this case where `ab == acr`, which will do both `scheduleOperandRead` and `scheduleLdMatrix`. Therefore, it makes no sense to do `scheduleLdMatrix` again for `acr` because `acr` is just `ab`.